### ISSUE TITLE
Lab 3 Comments

### ISSUE THREAD

**Issue by peterasujan** [cs10/bjc-r#193]
Please leave any comments on lab 3 here.

**Comment by xtitter** [comment]
A smattering of comments, some high level, some not:
- [ ] "primer: variables" is too big, too much.  Is there no way to start with conditionals and introduce variable when you need to?  'variables' need more than one page.  I'd start students with only "script variables" and "set" and "change".  Don't show the 'make a variable' button!  And, don't show 'show variable' and 'hide variable'...
- [ ] variables really needs to be useful for them to get it.  so, they need to have a problem they need to solve with a local variable in a block
- [ ] "why do we need".  I like this, mostly.  Although, I'm not sure the motivation of "needing to run every single one of their blocks" is right; rather, programs need to make "choices" sometimes.  dunno
- [ ] 'booleans'.  This would be a good place for variables, maybe?  _not_ a good place to introduce predicates, because they have never even seen reporters of any kind yet.    They definitely need to understand blocks that return values.  (x position) was kind of like this -- we called it a 'sensor' block maybe, in topic 1?  Or, brian probably changed that.
- [ ] introduce ask/answer in an earlier step, befrore thinking about conditionals.
- [ ] if and if else: you really don't need a script variable here, since answer is a variable also.  Sure would be nice to have the reporter-ask.  Maybe we can put it in the import tools?
- [ ] functions using if and if else:  they need a lot of help making reporters!  make the first one, maybe?  Lead them through the dialogs...
- [ ] 'final project' -- don't call it 'final project', just tamogaotchi.

**Comment by xtitter** [comment]
What about number guessing and rock/paper/scissors?  Definitely add those.

And the leap year one?  Pull everything you can from the old topic 2.

Maybe have number guessing be the final project for now; tamogaotchi is probably too involved.

**Comment by peterasujan** [comment]
> What about number guessing and rock/paper/scissors? Definitely add those.

Can do.

> And the leap year one? Pull everything you can from the old topic 2.

Already planning on it.

> Maybe have number guessing be the final project for now; tamogaotchi is probably too involved.

Sure.

**Comment by xtitter** [comment]
- [ ] Brian/Dan think a necessary example is programming <=  (I think this is in old topic 4).  Make this the first reporter they ever program, which will be a predicate.  So, go slow on it a bit (need to introduce the "report" block, etc).  Before this you'll need to introduce true/false, shape of the block, using built-in predicates, etc. (most of these you already have.
- [ ] 'answer' isn't a variable, its a reporter.  Brian wants this terminology exact.

**Comment by xtitter** [comment]
First thing to do -- today? -- is to finalize the sequence of steps.  Adding empty placeholders steps is fine, with title/type/

**Comment by peterasujan** [comment]
> Brian/Dan think a necessary example is programming <= (I think this is in old topic 4). Make this the first reporter they ever program, which will be a predicate. So, go slow on it a bit (need to introduce the "report" block, etc). Before this you'll need to introduce true/false, shape of the block, using built-in predicates, etc. (most of these you already have.

I like it.

**Comment by peterasujan** [comment]
> variables really needs to be useful for them to get it. so, they need to have a problem they need to solve with a local variable in a block

Fair enough. What about this?
    Make a (command) block that takes in a word and a number. 
    It should say the word joined together number times.
    E.g. block(hello, 3), the sprite says "hellohellohello"

**Comment by peterasujan** [comment]
> "primer: variables" is too big, too much. Is there no way to start with conditionals and introduce variable when you need to? 'variables' need more than one page. I'd start students with only "script variables" and "set" and "change". Don't show the 'make a variable' button! And, don't show 'show variable' and 'hide variable'...

Yes, I think I can work it in somewhere else, especially if I remove the unnecessary stuff. Are you at all attached to the variables quiz?

**Comment by xtitter** [comment]
In "predict the output", you'll need to include a primer on ABS and  MOD

These are pretty sweet looking, though.  There isn't a need for them to be custom blocks, though -- they could just be scripts in the scripting pane.  Not a big deal.

**Comment by xtitter** [comment]
in "blocks using if and if-else"
in would be great if you could include a buggy solution to one of these that the student needed to fix, rather than just having them write them.

for (letter grade ()), I'm not sure everyone will know the scale (in high school).  

(NOTE: this step is functional - don't worry about these additions, or put them in an html comment for now, until the rest of everything is done)

**Comment by xtitter** [comment]
"more complex boolean expressions"

really dense page.  You know it really is shouting for some class="quoteOrange" !!  Maybe in the discuss with your neighbor paragraph. (But, I don't know that this is a useful thing -- they don't know and/or/not yet, so how are they going to have an intuition on how to code them?  Maybe just a "Think about how you could express these statements as code with if/else blocks and simple predicates?  Convince yourself that you need something more..."   except that is dumbly worded.

Also, just set things off, pull images out and center them --- breaks up the long text.   For instance, the and/or/not blocks shouldn't be inline, but be centered.

Don't like the A, B headings on the tables...  hm...  maybe they could be a variable named A and a variable (i.e., the orange circle) called B, and then the "A and B" would be the blocks with the variables in the slots?  

Don't do the XOR thing, not right now.  too much to say it here, and Dan/Brian commented that XOR is way less successful than >= as their first custom predicate.

**Comment by xtitter** [comment]
I like the practice section.  You do need that page on how to write a reporter.

The "self-test: complex booleans" has script variables which haven't been introduced yet!  I think that needs to go into practice with predicates, and you'll need a whole page or two on that (one page of explanation about variables -- you had something already that could be converted?  Or, base it on the one from the old lab 2, where variables used to be introduced).  And, one page of practice with script vars and set.

Set is hard. There should be a lot of practice there, but that could go into lab 4 I guess.

**Comment by peterasujan** [comment]
> Don't like the A, B headings on the tables... hm... maybe they could be a variable named A and a variable (i.e., the orange circle) called B, and then the "A and B" would be the blocks with the variables in the slots? 

Yeah, I wanted to do that, just didn't have time yet. I'll try to get to it.

> Don't do the XOR thing, not right now. too much to say it here, and Dan/Brian commented that XOR is way less successful than >= as their first custom predicate. 

For sure. Should I just remove the sentence "Try writing it yourself!" or should I not mention xor at all? I think it's important to warn them that "or" is inclusive.

**Comment by peterasujan** [comment]
> really dense page. You know it really is shouting for some class="quoteOrange" !! Maybe in the discuss with your neighbor paragraph. (But, I don't know that this is a useful thing -- they don't know and/or/not yet, so how are they going to have an intuition on how to code them? Maybe just a "Think about how you could express these statements as code with if/else blocks and simple predicates? Convince yourself that you need something more..." except that is dumbly worded.

Yeah I can kill that sentence.
